libraries/chain_kv/include/b1/session/undo_stack.hpp

```diff
@@ -197,4 +215,123 @@ typename undo_stack<Session>::const_variant_type undo_stack<Session>::bottom() c
    return { *m_head, nullptr };
 }
 
+template <typename Session>
+void undo_stack<Session>::open() {
+   if (m_datadir.empty())
+      return;
+
+   if (!fc::is_directory(m_datadir))
+      fc::create_directories(m_datadir);
+
+   auto undo_stack_dat = m_datadir / undo_stack_filename;
+   if( fc::exists( undo_stack_dat ) ) {
+      try {
+         std::string content;
+         fc::read_file_contents( undo_stack_dat, content );
+
+         fc::datastream<const char*> ds( content.data(), content.size() );
+
+         // validate totem
+         uint32_t totem = 0;
+         fc::raw::unpack( ds, totem );
+         EOS_ASSERT( totem == undo_stack_magic_number, eosio::chain::chain_exception,
+                     "Undo stack data file '${filename}' has unexpected magic number: ${actual_totem}. Expected ${expected_totem}",
+                     ("filename", undo_stack_dat.generic_string())
+                     ("actual_totem", totem)
+                     ("expected_totem", undo_stack_magic_number)
+         );
+
+         // validate version
+         uint32_t version = 0;
+         fc::raw::unpack( ds, version );
+         EOS_ASSERT( version >= undo_stack_min_supported_version && version <= undo_stack_max_supported_version,
+                    eosio::chain::chain_exception,
+                    "Unsupported version of Undo stack data file '${filename}'. "
+                    "Undo stack data version is ${version} while code supports version(s) [${min},${max}]",
+                    ("filename", undo_stack_dat.generic_string())
+                    ("version", version)
+                    ("min", undo_stack_min_supported_version)
+                    ("max", undo_stack_max_supported_version)
+         );
+
+         int64_t rev; fc::raw::unpack( ds, rev );
+         revision(rev);
+
+         size_t num_sessions; fc::raw::unpack( ds, num_sessions );
+         for( size_t i = 0; i < num_sessions; ++i ) {
+            session_type session;
+            if (m_sessions.empty())
+               session.attach(*m_head);
+            else
+               session.attach(m_sessions.back());
+
+            size_t num_updated_keys; fc::raw::unpack( ds, num_updated_keys );
+            for( size_t j = 0; j < num_updated_keys; ++j ) {
+               shared_bytes_reflect raw_key; fc::raw::unpack( ds, raw_key );
```

Comment:
If you implement the above methods for `shared_bytes`, this becomes...
```c++
shared_bytes key;
ds >> key;

shared_bytes value;
ds >> value;
```

Reply:
Thanks for the suggestion! This simplifies the code. Changed.